llmfoundry/models/hf/hf_causal_lm.py

```diff
@@ -76,197 +114,181 @@ def __init__(self, om_model_config: Union[DictConfig,
+        if dist.get_local_rank() == 0:
+            with open(signal_file_path, 'wb') as f:
+                f.write(b'local_rank0_completed_download')
 
-            z_loss = om_model_config.get('z_loss', 0.0)
+        # Avoid the collective call until the local rank zero has finished trying to download the checkpoint
+        # so that we don't timeout for large downloads. This syncs all processes on the node
+        with dist.local_rank_zero_download_and_wait(signal_file_path):
+            # Then, wait to ensure every node has finished downloading the checkpoint
+            dist.barrier()
 
-            attention_patch_type = om_model_config.get('attention_patch_type',
-                                                       None)
-            if attention_patch_type is not None:
-                if model.config.model_type != 'llama':
-                    raise ValueError(
-                        f'attention_patch_type is only supported for llama models, but got {model.config.model_type}'
-                    )
+        if dist.get_local_rank() == 0:
+            os.remove(signal_file_path)
+
+        if attention_patch_type is not None:
+            self._patch_attention_type(model, attention_patch_type)
+
+        # Hugging Face's weight tying does not succeed if the model is inited on meta device
+        # so we manually apply the weight tying here
+        if model.config.tie_word_embeddings and resolved_init_device == 'meta':
+            model.tie_weights()
+
+        peft_config = None
+        if peft_config_dict is not None:
+            peft_config = self._get_peft_config(peft_config_dict)
+
+        super().__init__(
+            model=model,
+            shift_labels=True,
+            tokenizer=tokenizer,
+            metrics=train_metrics,
+            eval_metrics=eval_metrics,
+            z_loss=z_loss,
+            init_device=init_device,
+            peft_config=peft_config,
+        )
+
+    @staticmethod
+    def _patch_attention_type(model: PreTrainedModel,
+                              attention_patch_type: str) -> None:
+        if model.config.model_type != 'llama':
+            raise ValueError(
+                f'attention_patch_type is only supported for llama models, but got {model.config.model_type}'
+            )
 
-                log.debug(
-                    f'Patching llama attention with {attention_patch_type} attention'
+        log.debug(
+            f'Patching llama attention with {attention_patch_type} attention')
+        from transformers.models.llama.modeling_llama import LlamaAttention
+
+        from llmfoundry.models.layers.llama_attention_monkeypatch import \
+            get_llama_attention_patch_fn
+        LlamaAttention.forward = get_llama_attention_patch_fn(
+            attention_patch_type)
+        model.config.use_cache = False
+
+    @staticmethod
+    def _get_peft_config(peft_config_dict: Dict[str, Any]) -> 'PeftConfig':
+        if peft_installed:
```

Comment:
nit do early fail and maybe raise ImportError instead? 
```
        if not peft_installed:
              raise ImportError(...)
```

out of scope for PR, but it would be nice to have some error utiities like composer's `MissingConditionalImportError`